Sprint 12

KiMate34 opened the pull request:
Add SharedPreferences in SearchActivity, SettingsActivity
Created class SearchHistory

KiMate34 committed:
Add SharedPreferences in SearchActivity, SettingsActivity
Created class SearchHistory

Charmandik commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -23,17 +25,24 @@ class SearchActivity : AppCompatActivity() {
+        const val PLAYLIST_MAKER_PREFERENCES_TRACK = "playlist_maker_preferences"
     }
 
+    private lateinit var searchResultLayout: LinearLayout
```
Вместо использования lateinit, я рекомендую тебе использовать подход с добавлением nullability к типу объекта с оператором ?. и сразу проинициализировать его значением null. Это решение позволяет избежать ошибок, связанных с попытками доступа к неинициализированным объектам, которые могут привести к непредсказуемому поведению или крашам приложения. Хотя это может добавить некоторую дополнительную работу при обращении к объекту через оператор ?. , но это значительно повысит безопасность приложения, так как ты будешь уверен, что объект всегда либо имеет значение, либо равен null, и не будет вызывать исключительных ситуаций.Можешь еще прочитать статью на эту тему https://habr.com/ru/companies/maxilect/articles/590061

KiMate34 replied:
Благодарю за совет и за статью! Применил данную мудрость к своей программе)

KiMate34 replied:
Но тогда появляется множество (!!). Выглядит это не слишком надёжно) Как можно было бы сделать правильно?

Charmandik replied:
Наоборот, использовать !! не нужно, делай через оператор "?"

Charmandik commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -79,6 +109,14 @@ class SearchActivity : AppCompatActivity() {
+            searchHistory.clear()
+            historyTrackAdapter.tracks.clear()
+            historyTrackAdapter.notifyDataSetChanged()
+            textHistoryTitle.visibility = View.GONE
```
Для управления видимостью можно использовать свойство isVisible, устанавливая значение false для View.GONE и true для View.VISIBLE
Например, если нам нужно установить View.VISIBLE:
myViewName.isVisible = true
Или если нужно установить View.GONE, то : 
myViewName.isVisible = false

Возможно, тебе больше понравится такой подход

KiMate34 replied:
Забавно, что ты тоже за то, чтобы использовать такой способ установки визуализации) А подскажи пожалуйста, так в основном сейчас делают или может архитектурно так будет более правильно? Или это просто для читаемости кода делается?

Charmandik replied:
Для читаемости больше, если к примеру есть какое-то условие видимости view, то выглядит лучше ->
myView.isVisible = items.size > 10

Charmandik commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -87,6 +125,20 @@ class SearchActivity : AppCompatActivity() {
+            }
+        }
+
         val simpleTextWatcher = object : TextWatcher {
```
Функцию расширения doOnTextChanged можно использовать вместо класса TextWatcher для отслеживания изменений текста в EditText. Вот пример использования:

editText.doOnTextChanged { text, start, before, count ->
    // Здесь можно выполнять действия при изменении текста    // text - новый текст
    // start - начальная позиция изменения    // before - количество символов, удаленных из текста
    // count - количество добавленных символов в текст}

Метод doOnTextChanged принимает лямбда-выражение, которое будет вызываться при каждом изменении текста. Внутри этого выражения можно выполнять необходимые действия в соответствии с изменениями текста.

Charmandik commented on app/src/main/res/layout/activity_search.xml:
```diff
@@ -56,47 +56,83 @@
-            app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"
-            android:visibility="visible"/>
+            <LinearLayout
+                android:id="@+id/searchResultLayout"
```
Для более простого использования различных классов view рекомендую использовать следующий трюк: К имени переменной добавлять префикс в виде "аббревиатуры" класса, к которой она относится, для TextView - tv, ImageView - iv, RecyclerView - rv и т.д.. Например, у тебя есть RecyclerView для треков, в этом случае будет rvTracks

KiMate34 replied:
Понял! Применил этот трюк к своим айди view)

KiMate34 committed:
Change class SearchActivity and SettingsActivity
Change text id activity_search.xml, activity_track.xml
Create class App, added in AndroidManifest

Charmandik commented on app/src/main/java/com/example/playlistmaker/SettingsActivity.kt:
```diff
@@ -20,36 +20,32 @@ class SettingsActivity : AppCompatActivity() {
+            (applicationContext as App).switchTheme(isChecked)
         }
-        switchDarkTheme.isChecked = sharedPref.getBoolean(SWITCH_KEY, false)
+        switchDarkTheme!!.isChecked = sharedPref.getBoolean(SWITCH_KEY, false)
```
видимо я как-то плохо объяснил, но использовать оператор !! тоже достаточно плохая практика)
